File: app/src/main/java/com/mecong/myalarm/sleep_assistant/media_selection/OnlineMediaFragment.java
package com.mecong.myalarm.sleep_assistant.media_selection;

import android.app.Dialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.mecong.myalarm.R;
import com.mecong.myalarm.model.SQLiteDBHelper;
import com.mecong.myalarm.sleep_assistant.SleepAssistantViewModel;

import java.net.MalformedURLException;
import java.net.URL;

import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import lombok.experimental.FieldDefaults;

@FieldDefaults(level = AccessLevel.PRIVATE)
@NoArgsConstructor //required
public class OnlineMediaFragment extends Fragment implements MediaItemViewAdapter.ItemClickListener {

    MediaItemViewAdapter adapter;
    SleepAssistantViewModel model;


    OnlineMediaFragment(SleepAssistantViewModel model) {
        this.model = model;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_online_media, container, false);
    }

    private void addUrl(String url) {
        SQLiteDBHelper sqLiteDBHelper = SQLiteDBHelper.getInstance(this.getContext());
        sqLiteDBHelper.addMediaUrl(url);
        adapter.changeCursor(sqLiteDBHelper.getAllOnlineMedia());
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        final RecyclerView recyclerView = view.findViewById(R.id.mediaListView);
        SQLiteDBHelper sqLiteDBHelper = SQLiteDBHelper.getInstance(this.getContext());

        recyclerView.setLayoutManager(new LinearLayoutManager(view.getContext()));
        adapter = new MediaItemViewAdapter(view.getContext(),
                sqLiteDBHelper.getAllOnlineMedia(), this, true);

        recyclerView.setAdapter(adapter);

        final Button buttonAdd = view.findViewById(R.id.buttonAdd);
        buttonAdd.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final Dialog dialog = new Dialog(getContext());
                dialog.setContentView(R.layout.url_input_dialog);

                final EditText textUrl = dialog.findViewById(R.id.textUrl);
                Button buttonOk = dialog.findViewById(R.id.buttonOk);
                buttonOk.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        try {
                            new URL(textUrl.getText().toString());

                            addUrl(textUrl.getText().toString());
                            dialog.dismiss();
                        } catch (MalformedURLException mue) {
                            Toast.makeText(getContext(), "Url is not valid", Toast.LENGTH_SHORT).show();
                        }
                    }
                });

                final Button buttonCancel = dialog.findViewById(R.id.buttonCancel);
                buttonCancel.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        dialog.dismiss();
                    }
                });

                WindowManager.LayoutParams lp = new WindowManager.LayoutParams();
                lp.copyFrom(dialog.getWindow().getAttributes());
                lp.width = WindowManager.LayoutParams.MATCH_PARENT;
                lp.height = WindowManager.LayoutParams.WRAP_CONTENT;
                dialog.show();
                dialog.getWindow().setAttributes(lp);
            }
        });
    }


    @Override
    public void onItemClick(String url, int position) {
        SleepAssistantViewModel.PlayList newPlayList =
                new SleepAssistantViewModel.PlayList(url, url, SleepMediaType.ONLINE);
        model.setPlaylist(newPlayList);
    }

    @Override
    public void onItemDeleteClick(int position) {
        long itemId = adapter.getItemId(position);
        SQLiteDBHelper sqLiteDBHelper = SQLiteDBHelper.getInstance(this.getContext());
        sqLiteDBHelper.deleteOnlineMedia(String.valueOf(itemId));
        adapter.changeCursor(sqLiteDBHelper.getAllOnlineMedia());
    }
}

File: app/src/main/java/com/mecong/tenderalarm/sleep_assistant/media_selection/OnlineMediaFragment.kt
package com.mecong.tenderalarm.sleep_assistant.media_selection

import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.mecong.tenderalarm.R
import com.mecong.tenderalarm.databinding.FragmentOnlineMediaBinding
import com.mecong.tenderalarm.model.MediaEntity
import com.mecong.tenderalarm.model.PropertyName
import com.mecong.tenderalarm.model.SQLiteDBHelper.Companion.sqLiteDBHelper
import com.mecong.tenderalarm.sleep_assistant.Media
import com.mecong.tenderalarm.sleep_assistant.SleepAssistantPlayListActive
import com.mecong.tenderalarm.sleep_assistant.SleepAssistantPlayListIdle
import org.greenrobot.eventbus.EventBus
import java.net.MalformedURLException
import java.net.URL

class OnlineMediaFragment : Fragment(), FileItemClickListener {
  private var mediaItemViewAdapter: MediaItemViewAdapter? = null

  private var _binding: FragmentOnlineMediaBinding? = null

  // This property is only valid between onCreateView and onDestroyView.
  private val binding get() = _binding!!


  override fun onCreateView(
    inflater: LayoutInflater, container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View { // Inflate the layout for this fragment
    _binding = FragmentOnlineMediaBinding.inflate(inflater, container, false)
    return binding.root
  }

  private fun addUrl(title: String, url: String) {
    val sqLiteDBHelper = sqLiteDBHelper(this.requireContext())!!
    sqLiteDBHelper.addMediaUrl(title, url)
    mediaItemViewAdapter!!.updateDataSet(sqLiteDBHelper.allOnlineMedia)
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    val sqLiteDBHelper = sqLiteDBHelper(this.requireContext())!!
    binding.mediaListView.layoutManager = LinearLayoutManager(view.context)

    val list = sqLiteDBHelper.allOnlineMedia.use {
      generateSequence { if (it.moveToNext()) it else null }
        .map { MediaEntity.fromCursor(it) }
        .toList()
    }

    val savedActiveTab = sqLiteDBHelper.getPropertyInt(PropertyName.ACTIVE_TAB)

    var selectedPosition = -1
    if (savedActiveTab == 1) {
      selectedPosition = sqLiteDBHelper.getPropertyInt(PropertyName.TRACK_NUMBER) ?: 0
      initPlaylist(selectedPosition, false)
    }

    mediaItemViewAdapter = MediaItemViewAdapter(view.context, selectedPosition, list, this, true)

    binding.mediaListView.scrollToPosition(selectedPosition)

    binding.mediaListView.adapter = mediaItemViewAdapter

    binding.buttonAdd.setOnClickListener {
      val dialog = Dialog(requireContext(), R.style.UrlDialogCustom)
      dialog.setContentView(R.layout.url_input_dialog)
      val textUrl = dialog.findViewById<EditText>(R.id.textUrl)
      val titleUrl = dialog.findViewById<EditText>(R.id.textTitle)
      val buttonOk = dialog.findViewById<Button>(R.id.buttonOk)
      val buttonOkTop = dialog.findViewById<Button>(R.id.buttonOkTop)
      val addUrlListener: (v: View) -> Unit = {
        try {
          addUrl(titleUrl.text.toString(), URL(textUrl.text.toString()).toString())
          dialog.dismiss()
        } catch (mue: MalformedURLException) {
          Toast.makeText(
            context,
            requireContext().getString(R.string.url_invalid_warning),
            Toast.LENGTH_SHORT
          ).show()
        }
      }
      buttonOk.setOnClickListener(addUrlListener)
      buttonOkTop.setOnClickListener(addUrlListener)
      val buttonCancel = dialog.findViewById<Button>(R.id.buttonCancel)
      val buttonCancelTop = dialog.findViewById<Button>(R.id.buttonCancelTop)
      buttonCancel.setOnClickListener { dialog.dismiss() }
      buttonCancelTop.setOnClickListener { dialog.dismiss() }
      val lp = WindowManager.LayoutParams()
      lp.copyFrom(dialog.window!!.attributes)
      lp.width = WindowManager.LayoutParams.MATCH_PARENT
      lp.height = WindowManager.LayoutParams.WRAP_CONTENT
      dialog.show()
      dialog.window!!.attributes = lp
    }
  }

  override fun onFileItemClick(url: String?, position: Int) {
    initPlaylist(position, true)
  }

  private fun initPlaylist(position: Int, active: Boolean) {
    val sqLiteDBHelper = sqLiteDBHelper(this.requireContext())!!

    val list = sqLiteDBHelper.allOnlineMedia.use {
      generateSequence { if (it.moveToNext()) it else null }
        .map { MediaEntity.fromCursor(it) }
        .toList()
    }

    val media = list.map { Media(it.uri, it.header) }.toList()

    val newPlayList = if (active)
      SleepAssistantPlayListActive(position, media, SleepMediaType.ONLINE, -1)
    else
      SleepAssistantPlayListIdle(position, media, SleepMediaType.ONLINE, -1)

    EventBus.getDefault().postSticky(newPlayList)

  }

  override fun onFileItemDeleteClick(itemId: Int) {
    val sqLiteDBHelper = sqLiteDBHelper(this.requireContext())!!
    sqLiteDBHelper.deleteOnlineMedia(itemId.toString())
    mediaItemViewAdapter!!.updateDataSet(sqLiteDBHelper.allOnlineMedia)
  }

}